sync/server.md

```diff
@@ -0,0 +1,404 @@
+
+More information about the GraphQL terminology and schema can be found at https://graphql.org/learn/schema/
+
+## Server Technology
+
+The Sync Server will provide a GraphQL API for Apps to store and retrieve data from, based on a defined Data Schema.
+The server will be written in Javascript for Node.js 8. It can use any language features (ES6) that are available in that runtime. Babel or Typescript will not be used.
+The server will leverage various existing graphql libraries, such as `graphql`, `graphql-subscriptions`, `graphql-tools`, `apollo-server-express` and `subscriptions-transport-ws`.
+The server will use the standard module mechanism, `npm`, for installing and managing these packages.
+Npm `scripts` will be used extensively for running CI and local development tasks, executing cli tools directly e.g. `nodemon`, `standard`, `mocha`.
+Build tools such as `grunt` or `gulp` will *not* be used as they introduce extra overhead where a simple cli command will suffice. See https://www.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/ for more info.
+
+## Architecture
+
+![architecture](./architecture.png)
+src: https://docs.google.com/drawings/d/1OmMY1bbI8HDhlZ_cIirv7FYTEqRRV-Hh-g0ANtYEgBY/edit
+
+The Sync server is intended to be horizontally scalable.
+As such, it will us a [shared-nothing-architecture](https://en.wikipedia.org/wiki/Shared-nothing_architecture).
+In addition, the server will be developed following the [12 factor app](https://12factor.net/) pattern.
+The architecture will be kept simple, where appropriate.
+There will be a single server process which serves as the following:
+
+* Static file server for the Admin UI
+* API for managing the Data Schema and other GraphQL & Data Sync features
+* GraphQL API for Apps to send Queries & Mutations to
+* Subscriptions endpoint for Apps to connect to
+
+The main developer flow when using the Sync Server will be something like this:
+
+* Open the Data Sync Admin UI
+* Define the Data Schema
+* Configure Data Source(s)
+* Configure Resolvers for any Queries or Mutations
+* Update/modify the Data Schema as needed, continuously saving and trying out the Schema
+
+The last part of the flow is very important.
+This implies a fast feedback loop when developing your Data Schema.
+To ensure the feedback loop is as fast as possible, the architecture of sync will be slightly different that the proof of concept architecture.
+Both the backend for the Admin UI, and the GraphQL API that Apps talk to will be part of the 1 Data Sync server.
+This server will continously run, without any restart needed when the Data Schema changes.
+To facilitate this, the mounting of the graphql express middleware will need to be done in such a way that it always rebuilds the schema based on the latests available. In reality, the parsed schema should be cached, and only reloaded/reparsed if it has changed. Here is a sample graphql endpoint mounting in express:
+
```

Comment:
What's the idea here to make sure schema changes are picked up? 
Do we want to have an endpoint like "/updateSchema"? Or, something like store Data Schema, Resolver Mappings, Data Sources, Webhooks in configs/secrets and watch them using K8s mechanisms (like we do at push operator: https://github.com/aerogear/ups-config-operator/blob/a0be10d159752ad78a3897bd0645dfb7fa1d7502/pkg/configOperator/kubeHelper.go#L44)

Reply:
> What's the idea here to make sure schema changes are picked up?

Correct.

> Do we want to have an endpoint like "/updateSchema"? Or, something like store Data Schema, Resolver Mappings, Data Sources, Webhooks 

Yes

> in configs/secrets and watch them using K8s mechanisms (like we do at push operator: https://github.com/aerogear/ups-config-operator/blob/a0be10d159752ad78a3897bd0645dfb7fa1d7502/pkg/configOperator/kubeHelper.go#L44)

Probably not this bit. See comment https://github.com/aerogear/proposals/pull/40/files#r198116375

Reply:
+1